include/mfmg/dealii_operator.templates.hpp

```diff
@@ -148,6 +148,146 @@ DealIITrilinosMatrixOperator<VectorType>::build_range_vector() const
+// multiplication with the transpose of B, i.e. C = A * B^T
+//
+// Note that it is different from deal.II's SparseMatrix::Tmmult(C, B) which
+// performs C = A^T * B
+void matrix_transpose_matrix_multiply(
+    dealii::TrilinosWrappers::SparseMatrix &C,
+    dealii::TrilinosWrappers::SparseMatrix const &B,
+    dealii::TrilinosWrappers::SparseMatrix const &A)
+{
+  // C = A * B^T
+  // C_ij = A_ik * B^T_kj = A_ik * B_jk
+
+  std::vector<dealii::types::global_dof_index> i_indices;
+  A.locally_owned_range_indices().fill_index_vector(i_indices);
+  std::vector<dealii::types::global_dof_index> j_indices;
+  B.locally_owned_range_indices().fill_index_vector(j_indices);
+
+  int const global_n_rows = A.row_partitioner().NumGlobalElements();
+  int const global_n_columns = B.row_partitioner().NumGlobalElements();
+  for (int j = 0; j < global_n_columns; ++j)
+  {
+    auto const local_j = B.locally_owned_range_indices().index_within_set(j);
+
+    auto const src = extract_row(B, j);
+
+    std::remove_const<decltype(src)>::type dst(A.locally_owned_range_indices(),
+                                               A.get_mpi_communicator());
+    A.vmult(dst, src);
+
+    // NOTE: getting an error that the index set is not compressed when calling
+    // IndexSet::index_within_set() directly on dst.locally_owned_elements() so
+    // ended up making a copy and calling IndexSet::compress() on it.  Not sure
+    // it is the best solution but this will do for now.
+    auto index_set = dst.locally_owned_elements();
+    index_set.compress();
+    for (int i = 0; i < global_n_rows; ++i)
+    {
+      auto const local_i = index_set.index_within_set(i);
+      if (local_i != dealii::numbers::invalid_dof_index)
+      {
+        auto const value = dst[i];
+        if (std::abs(value) > 1e-14) // is that an appropriate epsilon?
+          C.set(i, j, value);
+      }
+    }
+  }
+  C.compress(dealii::VectorOperation::insert);
+}
+
+template <typename VectorType>
+std::shared_ptr<MatrixOperator<VectorType>>
+DealIITrilinosMatrixfreeOperator<VectorType>::multiply_transpose(
+    MatrixOperator<VectorType> const &operator_b) const
+{
+  // Downcast to TrilinosMatrixOperator
+  auto downcast_operator_b =
+      static_cast<DealIITrilinosMatrixOperator<VectorType> const &>(operator_b);
+
+  auto a = this->get_matrix();
+  auto b = downcast_operator_b.get_matrix();
+
+  auto c = std::make_shared<matrix_type>(a->locally_owned_range_indices(),
+                                         b->locally_owned_range_indices(),
+                                         a->get_mpi_communicator());
+  // a->mmult(*c, *b);
+
+  matrix_transpose_matrix_multiply(*c, *b, *a);
+
+  return std::make_shared<DealIITrilinosMatrixfreeOperator<VectorType>>(c);
+}
+
+template <typename VectorType>
+std::shared_ptr<MatrixOperator<VectorType>>
+DealIITrilinosMatrixfreeOperator<VectorType>::multiply(
+    MatrixOperator<VectorType> const &operator_b) const
```

Comment:
Actually not sure that `multiply()` should be defined in `DealIITrilinosMatrixfreeOperator`.  I needed it at some point when I was doing changes by increments...  If we remove it, let's not forget to put `final` specifier back for `DealIITrilinosMatrixOperator::multiply()`.

Reply:
If it's not necessary, let's remove it.